File: ios/Mynah/UI/Screens/HomeScreen.swift
import SwiftUI

/// Home: a board of work, in three states.
///
/// The first home was the composer — a blank box asking you to think of
/// something. The second was a feed, which answered "what has happened" when the
/// question people actually arrive with is "what is happening". This one is
/// sorted by state rather than by time: what is on a call now, what is waiting
/// for a person, what closed today. Everything older is one tap away under
/// History and does not compete for the top of the screen.
///
/// The bar at the foot carries the two destinations and, between them, the only
/// way to start a call — a button rather than a tab, because starting something
/// is not a place you navigate to and come back from.
struct HomeScreen: View {
    @ObservedObject var model: CallsViewModel
    let onOpenCall: (String) -> Void
    let onOpenDetail: (String) -> Void
    let onCompose: () -> Void
    let onComposeWith: (String) -> Void
    let onSearch: () -> Void
    let onUsage: () -> Void
    let onSettings: () -> Void
    let onHistory: () -> Void

    /// Only ticks while something is actually on a call — the elapsed time on
    /// the live card is the one thing here that changes on its own.
    @State private var now = Int(Date().timeIntervalSince1970 * 1000)

    private var live: Call? { model.calls.first(where: \.isLive) }

    /// Everything on a line right now. Plural on purpose — two calls can be
    /// running at once, and the old single-card top slot quietly hid the second.
    private var running: [Call] { model.calls.filter(\.isLive) }

    private var needsYou: [Call] {
        model.calls.filter { $0.status == CallStatus.transferring }
    }

    /// Closed since midnight. Older than that belongs to History: this section
    /// answers "did the thing I asked for this morning land", and a week of
    /// results underneath it makes that harder to see, not easier.
    private var doneToday: [Call] {
        let midnight = Calendar.current.startOfDay(for: Date()).timeIntervalSince1970 * 1000
        return model.calls.filter {
            !$0.isLive && $0.status != CallStatus.transferring && Double($0.createdAt) >= midnight
        }
    }

    /// Nothing to show and no way to fetch any: the whole page says so rather
    /// than an error card floating above an empty feed. Only when there is
    /// genuinely nothing — with calls already in hand, the error card over real
    /// history is more useful than blanking the screen.
    private var offline: Bool { model.error != nil && model.calls.isEmpty && !model.loading }

    var body: some View {
        VStack(spacing: 0) {
            topBar

            if offline {
                Offline {
                    model.clearError()
                    model.refresh()
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                board
            }

            // Offline is the one case where the middle button really is dead:
            // composing starts by asking the server to read the request, so
            // there is nothing behind it to reach.
            TaskTabBar(
                selected: .tasks,
                onSelect: { if $0 == .history { onHistory() } },
                onCompose: { if !offline { onCompose() } },
                composeEnabled: !offline
            )
        }
        .task {
            model.refresh(quiet: true)
            model.loadProfile()
            model.loadUsage()
            CallWatch.shared.requestPermission()
        }
        .task(id: live?.id) {
            while live != nil, !Task.isCancelled {
                now = Int(Date().timeIntervalSince1970 * 1000)
                try? await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }
    }

    private var topBar: some View {
        HStack(alignment: .top, spacing: 0) {
            VStack(alignment: .leading, spacing: 2) {
                Text(greeting).wise(Type.title).foregroundStyle(Ink.text)
                Text(summary).wise(Type.caption).foregroundStyle(Ink.mute)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            // What is left to spend, where the eye already is. It used to be
            // three screens away under Settings, which is where people looked
            // for it only after a call failed to go out.
            if model.usage.balance > 0 {
                Text(t("board_calls_left", model.usage.balance))
                    .wise(Type.chip)
                    .foregroundStyle(Ink.deep)
                    .padding(.horizontal, 11)
                    .padding(.vertical, 6)
                    .background(Ink.limePale, in: Capsule())
                    .contentShape(Capsule())
                    .onTapGesture(perform: onUsage)
            }

            IconCircle(icon: Wise.search, action: onSearch, size: 36, iconSize: 18)
                .accessibilityLabel(t("nav_search"))
            IconCircle(icon: Wise.gear, action: onSettings, size: 36)
                .accessibilityLabel(t("nav_settings"))
        }
        .padding(.leading, 20)
        .padding(.trailing, 12)
        .padding(.top, 12)
        .padding(.bottom, 4)
    }

    private var greeting: String {
        switch Calendar.current.component(.hour, from: Date()) {
        case ..<12: t("greeting_morning")
        case ..<18: t("greeting_afternoon")
        default: t("greeting_evening")
        }
    }

    /// The one line under the greeting. Says what is outstanding, or says
    /// plainly that nothing is — a screen that only ever describes activity
    /// leaves someone guessing whether an empty board means idle or broken.
    private var summary: String {
        var parts: [String] = []
        if !running.isEmpty { parts.append(t("board_summary_running", running.count)) }
        if !needsYou.isEmpty { parts.append(t("board_summary_needs_you", needsYou.count)) }
        return parts.isEmpty ? t("board_summary_idle") : parts.joined(separator: " · ")
    }

    private var board: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                if let error = model.error {
                    ErrorCard(message: error) { model.clearError() }
                    Spacer().frame(height: 14)
                }

                if running.isEmpty && needsYou.isEmpty && doneToday.isEmpty {
                    Spacer().frame(height: 24)
                    EmptyBoard(onExample: onComposeWith)
                }

                section(t("board_needs_you"), needsYou.isEmpty) {
                    ForEach(needsYou) { call in
                        NeedsYouRow(
                            call: call,
                            onTakeOver: { model.takeOver(call.id) },
                            onOpen: { onOpenCall(call.id) }
                        )
                        Spacer().frame(height: 10)
                    }
                }

                section(t("board_running"), running.isEmpty) {
                    ForEach(running) { call in
                        RunningCard(call: call, now: now) { onOpenCall(call.id) }
                        Spacer().frame(height: 10)
                    }
                }

                if !doneToday.isEmpty {
                    HStack(alignment: .firstTextBaseline) {
                        Text(t("board_done_today"))
                            .wise(Type.labelSmall)
                            .foregroundStyle(Ink.mute)
                        Spacer()
                        Text(t("board_all_history"))
                            .wise(Type.chip)
                            .foregroundStyle(Ink.deep)
                            .contentShape(Rectangle())
                            .onTapGesture(perform: onHistory)
                    }
                    .padding(.top, 18)
                    .padding(.bottom, 8)

                    GroupedCard {
                        ForEach(doneToday) { call in
                            FeedRow(
                                dot: call.presentation().dot,
                                title: call.headline(),
                                subtitle: call.subline(),
                                onTap: { onOpenDetail(call.id) },
                                right: {
                                    Text(formatClock(call.endedAt ?? call.createdAt))
                                        .wise(Type.mono)
                                        .foregroundStyle(Ink.mute)
                                    if let cost = call.cost {
                                        Text(costLabel(cost)).wise(Type.mono).foregroundStyle(Ink.mute)
                                    }
                                }
                            )
                        }
                    }
                }

                Spacer().frame(height: 24)
            }
            .padding(.horizontal, 20)
            .padding(.top, 4)
        }
        .refreshable { model.refresh(quiet: true) }
    }

    /// A labelled band, drawn only when it has something in it. An empty
    /// heading is a promise the screen does not keep.
    @ViewBuilder
    private func section<Content: View>(
        _ label: String,
        _ empty: Bool,
        @ViewBuilder content: () -> Content
    ) -> some View {
        if !empty {
            Text(label)
                .wise(Type.labelSmall)
                .foregroundStyle(Ink.mute)
                .padding(.top, 18)
                .padding(.bottom, 8)
            content()
        }
    }

}

/// The three lime bars. Small enough to draw rather than ship as an asset.
private struct Wordmark: View {
    var body: some View {
        HStack(alignment: .center, spacing: 1) {
            ForEach([8.0, 16.0, 10.0], id: \.self) { height in
                Capsule().fill(Ink.lime).frame(width: 4, height: height)
            }
        }
    }
}

/// The call happening right now. Black, so it reads as the one live thing on a
/// page of finished ones, and tapping anywhere on it goes to the call.
private struct Offline: View {
    let onRetry: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Icon(Wise.wifiOff, size: 22)
                .foregroundStyle(Ink.rim)
                .frame(width: 52, height: 52)
                .background(Ink.canvasSoft, in: Circle())
            Spacer().frame(height: 14)
            Text(t("offline_title")).wise(Type.section).foregroundStyle(Ink.text)
            Spacer().frame(height: 6)
            Text(t("offline_body"))
                .wise(Type.caption)
                .foregroundStyle(Ink.mute)
                .multilineTextAlignment(.center)
            Spacer().frame(height: 18)
            OutlineButton(
                label: t("action_retry"),
                action: onRetry,
                height: 46,
                style: Type.buttonSmall,
                leading: { Icon(Wise.rotate, size: 15) }
            )
            .fixedSize()
        }
        .padding(.horizontal, 40)
    }
}

/// Nothing running, nothing waiting, nothing closed today.
///
/// The picture is five bars of the brand's own palette, breathing at different
/// rates. Deliberately not an illustration of a phone or a person: this screen
/// is the first thing a new account sees, and a drawing of the thing the app
/// does invites the reader to decode it. Five bars decode to nothing, which
/// leaves the words to say what to do — and the words are the part that
/// actually answers "what now".
///
/// It is the same shape as the equaliser on a running call, slowed down. When
/// the board has something on it, those bars are moving because a line is open;
/// here they are the same object at rest.
private struct EmptyBoard: View {
    let onExample: (String) -> Void

    var body: some View {
        VStack(spacing: 0) {
            SoundSculpture()
                .frame(height: 196)
                .padding(.bottom, 28)

            Text(t("board_empty_title"))
                .wise(Type.section)
                .foregroundStyle(Ink.text)
                .multilineTextAlignment(.center)
            Spacer().frame(height: 6)
            Text(t("board_empty_body"))
                .wise(Type.caption)
                .foregroundStyle(Ink.mute)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: 300)

            Spacer().frame(height: 22)
            // Tapping one opens the composer with it already typed. An empty box
            // asking someone to think of something is the hardest screen in the
            // app; three real errands are a way past it.
            HStack(spacing: 8) {
                example(t("board_example_booking"), t("home_example_3"))
                example(t("board_example_reschedule"), t("home_example_1"))
                example(t("board_example_parcel"), t("home_example_2"))
            }
        }
        .frame(maxWidth: .infinity)
    }

    private func example(_ label: String, _ seed: String) -> some View {
        Text(label)
            .wise(Type.chip)
            .foregroundStyle(Ink.deep)
            .padding(.horizontal, 13)
            .padding(.vertical, 8)
            .background(Ink.limePale, in: Capsule())
            .contentShape(Capsule())
            .onTapGesture { onExample(seed) }
    }
}

/// Five capsules, five different rhythms.
///
/// The periods are deliberately coprime-ish — 4.4 to 5.8 seconds, each with its
/// own offset — so the group never lands back in step. A shared period reads as
/// a loading indicator, which is the one thing this must not be mistaken for:
/// nothing here is waiting on anything.
private struct SoundSculpture: View {
    private struct Bar { let height: CGFloat; let colour: Color; let period: Double; let delay: Double }

    private let bars: [Bar] = [
        Bar(height: 74,  colour: Ink.limePale,    period: 4.6, delay: 0),
        Bar(height: 132, colour: Ink.limeNeutral, period: 5.4, delay: 0.6),
        Bar(height: 176, colour: Ink.lime,        period: 5.0, delay: 0.2),
        Bar(height: 112, colour: Ink.text,        period: 5.8, delay: 0.9),
        Bar(height: 88,  colour: Ink.limeNeutral, period: 4.4, delay: 0.4),
    ]

    @State private var floating = false

    var body: some View {
        ZStack {
            HStack(spacing: 12) {
                ForEach(Array(bars.enumerated()), id: \.offset) { _, bar in
                    Capsule()
                        .fill(bar.colour)
                        .frame(width: 24, height: bar.height)
                        .offset(y: floating ? -7 : 7)
                        .animation(
                            .easeInOut(duration: bar.period / 2)
                                .repeatForever(autoreverses: true)
                                .delay(bar.delay),
                            value: floating
                        )
                }
            }

            // Two specks, off the grid. Without them the group sits dead centre
            // and reads as a chart; with them it reads as a composition.
            Circle().fill(Ink.rim).frame(width: 5, height: 5)
                .offset(x: -122, y: -78)
            Circle().fill(Ink.limeNeutral).frame(width: 4, height: 4)
                .offset(x: 104, y: 62)
        }
        .frame(maxWidth: .infinity)
        .onAppear { floating = true }
    }
}

File: android/app/src/main/java/com/voicecall/ui/screens/HomeScreen.kt
package com.voicecall.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBars
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.voicecall.R
import com.voicecall.data.Call
import com.voicecall.data.CallStatus
import com.voicecall.ui.CallsViewModel
import com.voicecall.ui.ErrorCard
import com.voicecall.ui.FeedRow
import com.voicecall.ui.GroupLabel
import com.voicecall.ui.OutlineButton
import com.voicecall.ui.GroupedCard
import com.voicecall.ui.headline
import com.voicecall.ui.presentation
import com.voicecall.ui.subline
import com.voicecall.ui.IconCircle
import com.voicecall.ui.PulsingDot
import com.voicecall.ui.StateChip
import com.voicecall.ui.Waveform
import com.voicecall.ui.clickableNoRipple
import com.voicecall.ui.costLabel
import com.voicecall.ui.dayLabel
import com.voicecall.ui.elapsedOf
import com.voicecall.ui.formatClock
import com.voicecall.ui.statusBarPadding
import com.voicecall.ui.theme.Ink
import com.voicecall.ui.theme.Type
import com.voicecall.ui.theme.Wise
import kotlinx.coroutines.delay

/** The three feed filters. Named apart from History's own, same package. */
private enum class FeedFilter(val label: Int) {
    ALL(R.string.filter_all),
    NEEDS_YOU(R.string.status_needs_you),
    DONE(R.string.status_done),
}

/**
 * Home: what the assistant has been doing.
 *
 * The old home was the composer — you arrived at a blank box and were asked to
 * think of something. This one opens on the work instead: whatever is on a call
 * right now at the top, then everything else grouped by day. Composing has moved
 * behind the bar at the foot, which is where it belongs once the app has any
 * history to show.
 *
 * There is no bottom navigation any more. Search, notifications and settings are
 * top-bar icons, and everything else is reached from a row.
 */
@Composable
fun HomeScreen(
    vm: CallsViewModel,
    onOpenCall: (String) -> Unit,
    onOpenDetail: (String) -> Unit,
    onCompose: () -> Unit,
    onSearch: () -> Unit,
    onNotifications: () -> Unit,
    onSettings: () -> Unit,
    onHistory: () -> Unit,
) {
    val state by vm.uiState.collectAsState()
    val profile by vm.profile.collectAsState()
    var filter by remember { mutableStateOf(FeedFilter.ALL) }
    val context = LocalContext.current

    LaunchedEffect(Unit) { vm.refresh(quiet = true); vm.loadProfile() }

    // Only ticks while something is actually on a call — the elapsed time on the
    // live card is the one thing here that changes on its own.
    var now by remember { mutableLongStateOf(System.currentTimeMillis()) }
    val anyLive = state.calls.any { it.isLive }
    LaunchedEffect(anyLive) {
        while (anyLive) {
            now = System.currentTimeMillis()
            delay(1000)
        }
    }

    val live = state.calls.firstOrNull { it.isLive }
    // The live call has its own card at the top, so it is not repeated in the
    // list below it.
    val rest = state.calls.filter { it.id != live?.id }.filter { call ->
        when (filter) {
            FeedFilter.ALL -> true
            FeedFilter.NEEDS_YOU -> call.status == CallStatus.TRANSFERRING
            FeedFilter.DONE -> call.status == CallStatus.COMPLETED
        }
    }

    Column(Modifier.fillMaxSize().padding(top = statusBarPadding())) {
        // --- top bar ---
        Row(
            Modifier.fillMaxWidth().padding(start = 20.dp, end = 12.dp, top = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(Modifier.weight(1f), verticalAlignment = Alignment.CenterVertically) {
                Wordmark()
                Spacer(Modifier.width(8.dp))
                Text(stringResource(R.string.app_name), style = Type.Wordmark, color = Ink.Text)
            }
            IconCircle(Wise.Search, stringResource(R.string.nav_search), onSearch, size = 38.dp, iconSize = 19.dp)
            IconCircle(
                Wise.Bell,
                stringResource(R.string.nav_notifications),
                onNotifications,
                size = 38.dp,
                badge = state.calls.any { it.status == CallStatus.TRANSFERRING },
            )
            IconCircle(Wise.Gear, stringResource(R.string.nav_settings), onSettings, size = 38.dp)
        }

        // Nothing to show and no way to fetch any: the whole page says so
        // rather than an error card floating above an empty feed. Only when
        // there is genuinely nothing — with calls already in hand, the error
        // card over real history is more useful than blanking the screen.
        val offline = state.error != null && state.calls.isEmpty() && !state.loading

        Box(Modifier.weight(1f)) {
            if (offline) {
                Offline(onRetry = { vm.clearError(); vm.refresh() })
            } else {
            Column(
                Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(start = 20.dp, end = 20.dp, top = 16.dp),
            ) {
                state.error?.let {
                    ErrorCard(it, vm::clearError)
                    Spacer(Modifier.height(14.dp))
                }

                if (live != null) {
                    LiveCallCard(live, now) { onOpenCall(live.id) }
                    Spacer(Modifier.height(18.dp))
                }

                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    FeedFilter.entries.forEach { f ->
                        StateChip(stringResource(f.label), filter == f) { filter = f }
                    }
                }
                Spacer(Modifier.height(12.dp))

                if (rest.isEmpty()) {
                    EmptyFeed()
                } else {
                    // Grouped by the day they were placed, newest day first, in
                    // the order the server already returned them.
                    rest.groupBy { dayLabel(context, it.createdAt, now) }.forEach { (day, calls) ->
                        GroupLabel(day, modifier = Modifier.padding(top = 4.dp, bottom = 8.dp))
                        GroupedCard(rows = calls.map { call ->
                            {
                                FeedRow(
                                    dot = call.presentation().dot,
                                    title = call.headline(),
                                    subtitle = call.subline(),
                                    pulsing = call.isLive,
                                    onClick = {
                                        if (call.isLive) onOpenCall(call.id) else onOpenDetail(call.id)
                                    },
                                    right = {
                                        Text(
                                            formatClock(call.endedAt ?: call.createdAt),
                                            style = Type.Mono,
                                            color = Ink.Mute,
                                        )
                                        call.cost?.let {
                                            Text(costLabel(it), style = Type.Mono, color = Ink.Mute)
                                        }
                                    },
                                )
                            }
                        })
                        Spacer(Modifier.height(16.dp))
                    }

                    Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                        Text(
                            stringResource(R.string.feed_view_all),
                            style = Type.Chip,
                            color = Ink.Deep,
                            modifier = Modifier.clickableNoRipple(onHistory).padding(vertical = 10.dp),
                        )
                    }
                }

                // Room for the composer bar to float over without covering the
                // last row.
                Spacer(Modifier.height(84.dp))
            }
            }

            // Still tappable when unverified — it leads to the composer, which
            // explains why and offers the way through. A dead bar would leave
            // someone staring at a hint with nothing to press.
            //
            // Offline is the one case where it really is dead: the composer's
            // first step is asking the server to read the request, so there is
            // nothing behind the bar to reach.
            ComposerBar(
                onClick = { if (!offline) onCompose() },
                blocked = profile?.canCall == false,
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .alpha(if (offline) 0.55f else 1f),
            )
        }
    }
}

/** The three lime bars. Small enough to draw rather than ship as an asset. */
@Composable
private fun Wordmark() {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(1.dp)) {
        listOf(8.dp, 16.dp, 10.dp).forEach { h ->
            Box(Modifier.width(4.dp).height(h).background(Ink.Lime, RoundedCornerShape(99.dp)))
        }
    }
}

/**
 * The call happening right now. Black, so it reads as the one live thing on a
 * page of finished ones, and tapping anywhere on it goes to the call.
 */
@Composable
private fun LiveCallCard(call: Call, now: Long, onClick: () -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(18.dp))
            .background(Ink.Text, RoundedCornerShape(18.dp))
            .clickableNoRipple(onClick)
            .padding(18.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            PulsingDot(Ink.Lime, size = 7.dp)
            Spacer(Modifier.width(8.dp))
            Text(stringResource(R.string.status_on_the_call), style = Type.LabelSmall, color = Ink.Lime)
            Spacer(Modifier.weight(1f))
            Text(elapsedOf(call, now), style = Type.MonoBody, color = Ink.OnDarkMute)
        }
        Spacer(Modifier.height(10.dp))
        Text(
            call.businessName.ifBlank { stringResource(R.string.status_dialing_now) },
            style = Type.ListTitle,
            color = Ink.OnDark,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        Text(
            call.goal,
            style = Type.RowSub,
            color = Ink.OnDarkMute,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        Spacer(Modifier.height(14.dp))
        Row(verticalAlignment = Alignment.CenterVertically) {
            Waveform(Modifier.height(18.dp), barCount = 4, periodMs = 900, staggerMs = 150)
            Spacer(Modifier.width(10.dp))
            Text(
                call.transcript.lastOrNull()?.text.orEmpty(),
                style = Type.RowSub,
                color = Ink.OnDarkBody,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f),
            )
            Spacer(Modifier.width(8.dp))
            Text(stringResource(R.string.feed_enter_call), style = Type.Chip, color = Ink.Lime)
        }
    }
}

/**
 * The way to start a call, floating over the foot of the feed. The gradient
 * behind it fades the list out rather than cutting it, so it reads as the page
 * continuing underneath.
 */
@Composable
private fun ComposerBar(onClick: () -> Unit, blocked: Boolean, modifier: Modifier = Modifier) {
    // The gradient runs under the system navigation bar, but the bar itself has
    // to sit above it — this floats over the list, so nothing else is going to
    // reserve that space.
    val bottomInset = WindowInsets.navigationBars.asPaddingValues().calculateBottomPadding()
    Box(
        modifier
            .fillMaxWidth()
            .background(Brush.verticalGradient(listOf(Color.Transparent, Ink.Canvas, Ink.Canvas)))
            .padding(start = 16.dp, end = 16.dp, top = 20.dp, bottom = 14.dp + bottomInset),
    ) {
        Row(
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(16.dp))
                .background(Ink.Card, RoundedCornerShape(16.dp))
                .border(1.dp, Ink.Hairline, RoundedCornerShape(16.dp))
                .clickableNoRipple(onClick)
                .padding(start = 16.dp, end = 8.dp, top = 8.dp, bottom = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                stringResource(
                    if (blocked) R.string.feed_composer_blocked else R.string.feed_composer_hint,
                ),
                style = Type.Body,
                color = if (blocked) Ink.WarningDeep else Ink.Mute,
                modifier = Modifier.weight(1f),
            )
            Spacer(Modifier.width(11.dp))
            Box(
                Modifier
                    .size(40.dp)
                    .background(if (blocked) Ink.Warning else Ink.Lime, RoundedCornerShape(12.dp)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    if (blocked) Wise.Person else Wise.Mic,
                    null,
                    tint = if (blocked) Ink.WarningInk else Ink.OnLime,
                    modifier = Modifier.size(18.dp),
                )
            }
        }
    }
}

@Composable
/**
 * The whole page, when there is nothing to show and no way to fetch any.
 *
 * Borrows the empty state's shape — grey ring, title, a line of explanation —
 * and adds the one thing an empty feed does not need: something to press. The
 * body says calls already running carry on, because that is the first worry of
 * anyone who loses signal while the assistant is mid-call, and it is true: the
 * call is between the server and the phone network, and this app is only
 * watching it.
 */
private fun Offline(onRetry: () -> Unit) {
    Column(
        Modifier.fillMaxSize().padding(horizontal = 40.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Box(
            Modifier.size(52.dp).background(Ink.CanvasSoft, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Icon(Wise.WifiOff, null, tint = Ink.Rim, modifier = Modifier.size(22.dp))
        }
        Spacer(Modifier.height(14.dp))
        Text(stringResource(R.string.offline_title), style = Type.Section, color = Ink.Text)
        Spacer(Modifier.height(6.dp))
        Text(
            stringResource(R.string.offline_body),
            style = Type.Caption,
            color = Ink.Mute,
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(18.dp))
        OutlineButton(
            label = stringResource(R.string.action_retry),
            onClick = onRetry,
            height = 46.dp,
            style = Type.ButtonSmall,
            leading = { Icon(Wise.Rotate, null, tint = Ink.Text, modifier = Modifier.size(15.dp)) },
        )
    }
}

@Composable
private fun EmptyFeed() {
    Column(Modifier.fillMaxWidth().padding(top = 40.dp), horizontalAlignment = Alignment.CenterHorizontally) {
        Box(Modifier.size(52.dp).background(Ink.CanvasSoft, CircleShape), contentAlignment = Alignment.Center) {
            Icon(Wise.Phone, null, tint = Ink.Rim, modifier = Modifier.size(22.dp))
        }
        Spacer(Modifier.height(14.dp))
        Text(stringResource(R.string.feed_empty_title), style = Type.Section, color = Ink.Text)
        Spacer(Modifier.height(6.dp))
        Text(stringResource(R.string.feed_empty_body), style = Type.Caption, color = Ink.Mute)
    }
}


